src/Controllers/RegistrationController.php

```diff
@@ -0,0 +1,268 @@
+            ]);
+        }
+        if ($this->config->get('enable_dect')) {
+            $rules = array_merge($rules, ['dect'  => 'optional|length:0:40']);
+        }
+        if ($this->config->get('enable_user_name')) {
+            $rules = array_merge($rules, [
+                'firstname'   => 'optional|alpha',
+                'lastname'  => 'optional|alpha',
+            ]);
+        }
+
+        $data = $this->validate($request, $rules);
+
+        if (User::whereName($data['login'])->count() > 0) {
+            $this->session->set('errors', array_merge($this->session->get('errors', []), ['registration.exists.login']));
+
+            return $this->showForm();
+        }
+
+        if (User::whereEmail($data['email'])->first()) {
+            $this->session->set('errors', array_merge($this->session->get('errors', []), ['registration.exists.email']));
+
+            return $this->showForm();
+        }
+
+        if (!$isPending && $data['password'] !== $data['password_confirmation']) {
+            $this->session->set('errors', array_merge($this->session->get('errors', []), ['validation.password.confirmed']));
+
+            return $this->showForm();
+        }
+
+        $user = new User([
+            'name'          => $data['login'],
+            'password'      => '',
+            'email'         => $data['email'],
+            'api_key'       => '',
+            'last_login_at' => null,
+        ]);
+        $user->save();
+
+        $contact = new Contact([
+            'dect'      => $data['dect'] ?? '',
+            'mobile'    => $data['mobile'] ?? '',
+        ]);
+        $contact->user()
+            ->associate($user)
+            ->save();
+
+        $personalData = new PersonalData([
+            'first_name'            => $data['firstname'] ?? '',
+            'last_name'             => $data['lastname'] ?? '',
+            'shirt_size'            => $data['tshirt_size'] ?? '',
+            'planned_arrival_date'  => isset($data['planned_arrival_date'])
+                ? Carbon::createFromFormat('Y-m-d', $data['planned_arrival_date'])
+                : null,
+        ]);
+        $personalData->user()
+            ->associate($user)
+            ->save();
+
+        $settings = new Settings([
+            'language'          => $this->session->get('locale'),
+            'theme'             => $this->config->get('theme'),
+            'email_human'       => $request->has('email_human'),
+            'email_shiftinfo'   => $request->has('email_shiftinfo'),
+        ]);
+        $settings->user()
+            ->associate($user)
+            ->save();
+
+        $state = new State([]);
+        if ($this->config->get('autoarrive')) {
+            $state->arrived = true;
+            $state->arrival_date = new Carbon();
+        }
+        $state->user()
+            ->associate($user)
+            ->save();
+
+        DB::insert('INSERT INTO `UserGroups` (`uid`, `group_id`) VALUES (?, -20)', [$user->id]);
```

Comment:
This is not testable behavior and thus direct queries have to be avoided, esp. es it makes this controller only testable against a real Database and not only a sqlite in memory one.

And another point is that database queries in a controller are a no go...